### DISCUSSION TITLE
How does V lang compile to C?

### DISCUSSION THREAD

**Discussion by Kostia06** [vlang/v#17308] [General]
Hi everyone, I hope y'all doing good. I am currently working on my own programming language and I am curious how does V compile to C. Does it just translates V code into C code and compile C code? Or does it have another way of compile V code? Also, I understand that V is written in itself like Go and Rust, but how does that even work😅?

**Comment by JalonSolov** [comment]
V, as most compilers these days, parses the source into an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree).  Once the AST is created, the C backend (`vlib/v/gen/c`) walks the AST and output C code to a temporary file.  When that step it finished, it calls the C compiler to generate the binary.  This is the current default.

How it works is that it was first [bootstrapped](https://en.wikipedia.org/wiki/Bootstrapping_(compilers)) in Go, until it could compile itself.  Bootstrapping is necessary, of course, because you can't compile anything in a language that doesn't exist, yet.  You'll need to do the same with your language if you want it to compile itself.

**Reply by Kostia06** [reply]
Thank you! So, technically speaking V translate V code into C code into a file and compiles the C code using C compiler. Also, another question when the program compiles the temporary C file, does it use the compiler that is installed on the computer already or does V compiles it?

**Reply by JalonSolov** [reply]
It uses tcc by default, which the V Makefile installs for you under `thirdparty/tcc` mainly because it is _fast_, but it doesn't do optimization.  If you use the `-prod` switch, which turns on optimization, it uses what you have installed.  V is not a C compiler, so it can't compile the code itself.

Note that this is only with the default C backend.  V also has other backends in various states of usefulness, including JavaScript, WASM, and native (which does direct binary output to an exe file - but is, unfortunately, the least advanced at this time).  The native backend will become the default once it is more complete.  Someone has even been working on a FreePascal backend.
